modules/jixieBidAdapter.js

```diff
@@ -0,0 +1,281 @@
+        nv[0] = nv[0].replace(/^ /, '');
+        coo[decodeURIComponent(nv[0])] = decodeURIComponent(nv[1]);
+      }
+    }
+  } catch (err) {}
+  let ret = {};
+  cnames.forEach(function(oneC) {
+    if (coo[oneC]) {
+      ret[oneC] = coo[oneC];
+    }
+  });
+  return ret;
+}
+
+function setIds_(clientId, sessionId) {
+  let dd = null;
+  try {
+    dd = window.location.hostname.match(/[^.]*\.[^.]{2,3}(?:\.[^.]{2,3})?$/mg);
+  } catch (err1) {}
+  try {
+    internal.setCookie('_jx', clientId, dd,
+      new Date(new Date().setFullYear(new Date().getFullYear() + 1)));
+    internal.setCookie('_jxs', sessionId, dd,
+      new Date(new Date().setMinutes(new Date().getMinutes() + sidTTLMins_)));
+    storage.setDataInLocalStorage('_jx', clientId);
+    storage.setDataInLocalStorage('_jxs', sessionId);
+  } catch (error) {}
+}
+
+function fetchIds_() {
+  let ret = {
+    client_id_c: '',
+    client_id_ls: '',
+    session_id_c: '',
+    session_id_ls: ''
+  };
+  try {
+    let cookies = internal.getCookies(['_jx', '_jxs']);
+    let tmp = cookies['_jx'];
+    if (tmp) ret.client_id_c = tmp;
+    tmp = cookies['_jxs'];
+    if (tmp) ret.session_id_c = tmp;
+
+    tmp = storage.getDataFromLocalStorage('_jx');
+    if (tmp) ret.client_id_ls = tmp;
+    tmp = storage.getDataFromLocalStorage('_jxs');
+    if (tmp) ret.session_id_ls = tmp;
+  } catch (error) {}
+  return ret;
+}
+
+function getDevice_() {
+  return ((/(ios|ipod|ipad|iphone|android|blackberry|iemobile|opera mini|webos)/i).test(navigator.userAgent)
+    ? 'mobile' : 'desktop');
+}
+
+function pingTracking_(endpointOverride, qpobj) {
+  internal.ajax((endpointOverride || EVENTS_URL), null, qpobj, {
+    withCredentials: true,
+    method: 'GET',
+    crossOrigin: true
+  });
+}
+
+function jxOutstreamRender_(bidAd) {
+  bidAd.renderer.push(() => {
+    window.JixieOutstreamVideo.init({
+      sizes: [bidAd.width, bidAd.height],
+      width: bidAd.width,
+      height: bidAd.height,
+      targetId: bidAd.adUnitCode,
+      adResponse: bidAd.adResponse
+    });
+  });
+}
+
+function createRenderer_(bidAd, scriptUrl, createFcn) {
+  const renderer = Renderer.install({
+    id: bidAd.adUnitCode,
+    url: scriptUrl,
+    loaded: false,
+    config: {'player_height': bidAd.height, 'player_width': bidAd.width},
+    adUnitCode: bidAd.adUnitCode
+  });
+  try {
+    renderer.setRender(createFcn);
+  } catch (err) {
+    utils.logWarn('Prebid Error calling setRender on renderer', err);
+  }
+  return renderer;
+}
+
+function getMiscDims_() {
+  let ret = {
+    pageurl: '',
+    domain: '',
+    device: 'unknown'
+  }
+  try {
+    let refererInfo_ = getRefererInfo();
+    let url_ = ((refererInfo_ && refererInfo_.referer) ? refererInfo_.referer : window.location.href);
+    ret.pageurl = url_;
+    ret.domain = utils.parseUrl(url_).host;
+    ret.device = getDevice_();
+  } catch (error) {}
+  return ret;
+}
+
+// easier for replacement in the unit test
+export const internal = {
```

Comment:
@jxdeveloper1 hi there, I see that some of the functions that are in this `internal` object are alredy included in prebid. I suggest you to use the storageManager for getting and setting cookies.

Reply:
Hi there!
1. yes noted.
2. Ya actually thought about that for a long time (yes have been aware of the storage manager)
3. coz the cookie set and getting stuff is something our jixie scripts do in a few other
context (unrelated to header bidding) - pre-existed.
4. So here, I just want to do _exactly_ what the other scripts had been doing it
5. That's why after long agony, i did it like what I had committed.
6. However, if you still don't buy it, then yes I will try to adapt as you mentioned. Let me know please?
Thanks!
Renee

Reply:
We really prefer to move this code to use storageManager; in the past we've seen tests from other bidders starting to fail because we enhanced the logic of storageManager (that must follow all privacy rules regarding setting cookies and local storage) and we want to avoid this.

Reply:
Yup, noted with thanks. Will work on that!

Just want to have a little idea of the timing... Assuming I get this in Tue / Wed and *assuming* no other issues,
when approximately can jixie adaptor appear in an official prebid release? (Need to alert my boss if need to expect any long wait)
Thanks much!

Reply:
Usually there's a prebid release every friday, so we must push this by next friday.

Reply:
For the record Prebid.js releases are normally Tues or Weds. Please fix the unit tests to use storageManager - we believe tests that don't use this cause PBJS build failures.